File: soln-sliding-window-problems/maximum-number-of-vowels-in-a-substring-of-given-length.java
/**
 @author Farheen Bano
  
 Reference-
 https://leetcode.com/problems/maximum-number-of-vowels-in-a-substring-of-given-length/
*/

class Solution {
    public int maxVowels(String s, int k) {
        int i=0;
        int j=0;
        int count=0;
        int max=0;
        while(j<s.length()){
            char ch=s.charAt(j);
            if(isVowel(ch))
                count++;                    
            
            if(j-i+1==k){                
                max=Math.max(max,count);
                if(isVowel(s.charAt(i))){
                    count--;                        
                }                
                i++;
            }
            j++;
        }
        return max;
    }
    
    public boolean isVowel(char ch){
        if(ch=='a' || ch=='e' || ch=='i' || ch=='o' ||  ch=='u')
            return true;
        return false;
    }
}

File: Python/soln-sliding-window-problems/maximum-number-of-vowels-in-a-substring-of-given-length.py
class Solution:
    def maxVowels(self, s: str, k: int) -> int:
        vowels={'a':0, 'e':0, 'i':0, 'o':0, 'u':0}
        i,j=0,0
        n=len(s)
        max_sum=0
        
        while j<n:
            if s[j] in vowels:
                vowels[s[j]]+=1
            
            if j-i+1<k:
                j+=1
                
            elif j-i+1==k:
                count=sum(vowels.values())
                max_sum=max(max_sum,count)
                if s[i] in vowels:
                    vowels[s[i]]-=1
                
                j+=1 
                i+=1
                
        return max_sum
            